### PR TITLE
[PT FE] aten::cat input [static] type alignment via ConvertPromoteTypes

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - aten::cat inputs **static** type alignment via ConvertPromoteTypes
 - This PR introduce proposal of inserting ConvertPromoteTypes recursively for PT FE torch::cat inputs with static types
 - As incompatibility of static types leads to ov Concat error, this change is backward compatible
 - Support for dynamic types can be enabled with small adjustment of the conditions, but as it can affect existing models, so has been proposed within a separate PR:
     -  https://github.com/openvinotoolkit/openvino/pull/25077
 - Tests with `trace_model=True` (static types)

### Tickets:
 - 142388/143028

### CODE REVIEW

File: src/frontends/pytorch/src/op/cat.cpp [hunk 1]
```diff
@@ -56,6 +58,38 @@ OutputVector translate_cat_common(const NodeContext& context,
...
         return {context.mark_node(std::make_shared<v1::Reshape>(tensor, new_shape, false))};
     }
+
+    const auto first_in_type = list_elems.front().get_element_type();
+    const bool is_mixed_type =
+        list_elems.size() > 1 && (std::any_of(std::next(list_elems.begin()),
+                                              list_elems.end(),
+                                              [&first_in_type](const ov::Output<ov::Node>& input) {
+                                                  return input.get_element_type() != first_in_type ||
+                                                         input.get_element_type() == ov::element::dynamic;
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/cat.cpp:68]: If dynamic type is among other types we need to assume the worst case - when this type will be different from others, so please change that

File: src/frontends/pytorch/src/op/cat.cpp [hunk 2]
```diff
@@ -56,6 +58,38 @@ OutputVector translate_cat_common(const NodeContext& context,
...
+                                              }));
+    if (is_mixed_type) {
+        auto node_of_type = list_elems[0];
+        for (size_t i = 1; i < list_elems.size(); ++i) {
+            node_of_type =
+                std::make_shared<ov::op::v14::ConvertPromoteTypes>(node_of_type, list_elems[i], true)->output(0);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/cat.cpp:74]:
```suggestion
                std::make_shared<v14::ConvertPromoteTypes>(node_of_type, list_elems[i], true)->output(0);
```

File: src/frontends/pytorch/src/op/cat.cpp [hunk 3]
```diff
@@ -56,6 +58,38 @@ OutputVector translate_cat_common(const NodeContext& context,
...
+            context.mark_node(node_of_type.get_node_shared_ptr());
+        }
+
+        const auto unified_type = node_of_type.get_element_type();
+        auto inputs_vec = OutputVector(list_elems.begin(), list_elems.end());
+        for (size_t i = 0; i < inputs_vec.size(); ++i) {
+            context.mark_node(inputs_vec[i].get_node_shared_ptr());
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/cat.cpp:81]:
```suggestion
```
`mark_node` should only be called for new nodes, here you create no nodes

File: src/frontends/pytorch/src/op/cat.cpp [hunk 4]
```diff
@@ -56,6 +58,38 @@ OutputVector translate_cat_common(const NodeContext& context,
...
+            if (inputs_vec[i].get_element_type() != unified_type ||
+                inputs_vec[i].get_element_type() == ov::element::dynamic) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/cat.cpp:83]: Same, for dynamic type always assume worst case, when type will be different

**Review by mvafin — CHANGES_REQUESTED** [review]

File: tests/layer_tests/pytorch_tests/test_cat.py
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_cat.py:1]: please add more test cases. consider mixing python scalar types with tensors. Also add tests for cases [float32,int32,float32] and [int32,float32,int32]

File: src/frontends/pytorch/src/op/cat.cpp (hunk 2 above)
Comment (mitruska) [reply src/frontends/pytorch/src/op/cat.cpp]: Done

File: src/frontends/pytorch/src/op/cat.cpp (hunk 3 above)
Comment (mitruska) [reply src/frontends/pytorch/src/op/cat.cpp]: Okay, removed.

File: tests/layer_tests/pytorch_tests/test_cat.py
Comment (mitruska) [reply tests/layer_tests/pytorch_tests/test_cat.py:1]: Sure, I've added more tests. 
The scalar types are not allowed as inputs for torch.cat and ov Concat op.

**Comment by mitruska** [comment]
To provide the changes with possible impact on the supported models separately, 
I've created the following PR, extending the conditions for dynamic types case:
- https://github.com/openvinotoolkit/openvino/pull/25077

**Comment by mitruska** [comment]
Discussed with @mvafin and agreed to merge this feature for static and dynamic at once within:
- https://github.com/openvinotoolkit/openvino/pull/25077

**Pull request closed without merging**